app/workers/push_notifications/cleanup_worker.rb

```diff
@@ -0,0 +1,21 @@
+module PushNotifications
+  class CleanupWorker
+    include Sidekiq::Worker
+
+    sidekiq_options queue: :low_priority, retry: 10
+
+    def perform
+      redis = Redis.new(url: ENV["REDIS_RPUSH_URL"] || ENV["REDIS_URL"])
+      cursor = 0
+      until (cursor, keys = redis.scan(cursor, match: "rpush:notifications:*")).first.to_i.zero?
+        keys.each do |key|
+          next unless redis.ttl(key) == -1
+          next unless redis.type(key) == "hash"
+          next if redis.hget(key, "delivered").nil?
+
+          redis.expire(key, 8.hours.to_i)
```

Comment:
@rhymes Even though you mentioned this is premature and we're talking suuuuuper small scale here, it's a great opportunity to talk about the tradeoffs involved in optimizations, so bear with me on this. 😄 

Pipelining is mainly good for side effects rather than returning values ("I need to set this key, then expire that key, then push an item into a list in the other key"). When you're using them to return values, it _must_ run all of the queries through Redis, so while it's an optimization for latency specifically, it comes with the cost of potentially unneeded queries. That is, if the `ttl` query returns `-1`, we still ran the `type` and `hget` queries because we sent all 3 before checking a single result. When it comes to Redis, the cost of 2 unnecessary queries is pretty low, but you get the idea. 😃 

Ultimately, we're making 3 redis queries which, from my benchmarks in a prod console, have a median time of ~1.1ms of wall-clock time with less CPU time spent than Ruby can even check (it's coming up as 0 seconds of CPU time). That wall-clock time means the 3 combined queries in the guard clauses all happen faster than your monitor could display the first one. 😄 

Another cost of pipelining queries is that the code becomes more difficult to grok. Setting up variables involves either messing with lexical scoping confusion (since this is inside a worker, using ivars doesn't make much sense, so we'd probably end up using local vars) or destructuring assignment:

```ruby
# local vars working around lexical scoping
ttl = nil
type = nil
delivered = nil
redis.pipelined do |pipe|
  ttl = pipe.ttl(key)
  type = pipe.type(key)
  delivered = pipe.hget(key, "delivered")
end

next unless ttl == -1
next unless type == "hash"
next if delivered.nil?
```

```ruby
# destructuring assignment since `Redis#pipelined` returns an array
# containing all results of queries executed within the block
ttl, type, delivered = redis.pipelined do |pipe|
  ttl = pipe.ttl(key)
  type = pipe.type(key)
  delivered = pipe.hget(key, "delivered")
end

next unless ttl == -1
next unless type == "hash"
next if delivered.nil?
```

Since neither of these are as easy to read as 3 lines of sequential Redis queries, the latency of those queries would need to be really bad to justify this. :joy:

Reply:
Thank you for the great context!